# Cleaning Data

We can use Pandas to clean our data so that it is ready for analysis. Three different types of data issues that we will deal with are:
1. Missing Data
2. Duplicate Data
3. Incorrect Data

To take care of these data issues, we are able to use built in pandas functions to clean it efficiently. Before starting to clean the above, we can use the `info()` function to give us a snapshot of the type of data that is in our DataFrame and systematically change data into proper data types.

In [7]:
!jt -t chesterish -N

In [8]:
!jt -t chesterish -kl

In [1]:
import pandas
import numpy

dataframe_3 = pandas.read_csv('Data Source.csv')
dataframe_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1887 entries, 0 to 1886
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Store       1887 non-null   int64  
 1   Date        1887 non-null   object 
 2   CategoryID  1887 non-null   int64  
 3   ProductID   1887 non-null   object 
 4   Q1          1886 non-null   float64
 5   Q2          1722 non-null   object 
 6   Q3          1738 non-null   float64
 7   Q4          1719 non-null   float64
dtypes: float64(3), int64(2), object(3)
memory usage: 118.1+ KB


We can first start by seeing that the `Date` column is an `object` type, whereas the correct type for a date is `datetime`. We can correct this by inputting the follow code.

In [14]:
pandas.to_datetime(dataframe_3['Date'])

0      2019-03-10
1      2019-06-04
2      2019-01-16
3      2019-10-18
4      2019-04-11
          ...    
1882   2019-03-24
1883   2019-03-24
1884   2019-03-24
1885   2019-09-30
1886   2019-06-14
Name: Date, Length: 1887, dtype: datetime64[ns]

In [15]:
# In order to make a change to the DataFrame, the changes must be reinstantiated to overwrite with the changes.
dataframe_3['Date'] = pandas.to_datetime(dataframe_3['Date'])

dataframe_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1887 entries, 0 to 1886
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Store       1887 non-null   int64         
 1   Date        1887 non-null   datetime64[ns]
 2   CategoryID  1887 non-null   int64         
 3   ProductID   1887 non-null   object        
 4   Q1          1886 non-null   float64       
 5   Q2          1722 non-null   object        
 6   Q3          1738 non-null   float64       
 7   Q4          1719 non-null   float64       
dtypes: datetime64[ns](1), float64(3), int64(2), object(2)
memory usage: 118.1+ KB


## Missing Data

We can also see that from the `info()` function, there are null values in our data from looking at the Non-Null Count column. We can see some columns have a different non-null count, which means that these columns have more null values. To see in greater detail which columns and rows have null values, we can use the `isna()` function.

In [13]:
dataframe_3.isna()

Store   Date  CategoryID  ProductID     Q1     Q2     Q3     Q4
0     False  False       False      False  False  False  False   True
1     False  False       False      False  False  False  False   True
2     False  False       False      False  False  False  False   True
3     False  False       False      False  False  False  False   True
4     False  False       False      False  False  False  False   True
...     ...    ...         ...        ...    ...    ...    ...    ...
1882  False  False       False      False  False  False  False  False
1883  False  False       False      False  False  False  False  False
1884  False  False       False      False  False  False  False  False
1885  False  False       False      False  False   True   True  False
1886  False  False       False      False  False  False  False  False

[1887 rows x 8 columns]

This output can be difficult to interpret and action. A way to simplify this information is to pair `isna()` with the `sum()` function. Now, we can see the number of null values per each column

In [17]:
dataframe_3.isna().sum()

Store           0
Date            0
CategoryID      0
ProductID       0
Q1              1
Q2            165
Q3            149
Q4            168
dtype: int64

We will look at two different ways to deal with missing data:

 1. Removing null values from DataFrame rows and/or columns
 2. Filling null values with a constant or other values from the DataFrame

### Removing null values

To remove columns or rows that contain null values, we can use the `dropna()` function. We can start by dropping the rows where at least one element is null in the row. This reduces our DataFrame from 1887 rows to 1684 rows.

In [15]:
dataframe_3.dropna()

Store       Date  CategoryID ProductID        Q1        Q2        Q3  \
8         1 2019-06-17           9    19XXRP  16930.99  16562.49  15880.85   
9         1 2019-03-03          10   110XXRP  30721.50  31494.77  29634.13   
10        1 2019-03-12          11   111XXRP  24213.18  21760.75  18706.21   
11        1 2019-08-27          12   112XXRP   8449.54   8654.07   9165.98   
12        1 2019-04-22          13   113XXRP  41969.29   36476.4  37857.68   
...     ...        ...         ...       ...       ...       ...       ...   
1881     40 2019-03-24          44  4044XXRP   3280.00   3556.22   2965.33   
1882     40 2019-03-24          44  4044XXRP   3280.00   3556.22   2965.33   
1883     40 2019-03-24          44  4044XXRP   3280.00   3556.22   2965.33   
1884     40 2019-03-24          44  4044XXRP   3280.00   3556.22   2965.33   
1886     40 2019-06-14          43  4093XXRP  31080.68  26600.39  23709.23   

            Q4  
8     15175.52  
9     27921.96  
10    17306.61  
11     9015.37  
12    37467.32  
...        ...  
1881   3041.41  
1882   3041.41  
1883   3041.41  
1884   3041.41  
1886  22879.37  

[1684 rows x 8 columns]

We can confirm that rows have been dropped, as the number of rows compared to the initial DataFrame has decreased. We can also drop columns that have null values rather than rows with the same function, but specifying the axis.



In [22]:
dataframe_3.dropna(axis = 1)

Store       Date  CategoryID ProductID
0         1 2019-03-10           1  11XXRP-P
1         1 2019-06-04           2  12XXRP-Q
2         1 2019-01-16           3    13XXRP
3         1 2019-10-18           4    14XXRP
4         1 2019-04-11           5  15XXRP-A
...     ...        ...         ...       ...
1882     40 2019-03-24          44  4044XXRP
1883     40 2019-03-24          44  4044XXRP
1884     40 2019-03-24          44  4044XXRP
1885     40 2019-09-30          49  4049XXRP
1886     40 2019-06-14          43  4093XXRP

[1887 rows x 4 columns]

We see that the columns that have any null values have been dropped. 

### Filling null values

However, completely dropping columns and rows that have null values is usually not an efficient method of cleaning data, as lots of data can be lost. 

A more common approach to cleaning these null values is to fill them in using the `fillna()` function. Two popular uses of the function is:

 1. Fill the null values with 0 or any value
 2. Fill the null values based on the next/previous number
 
With these methods, we can still retain valuable data and not drop the entire column/row.

In [23]:
# filling NaNs with 0
dataframe_3.fillna(0)

Store       Date  CategoryID ProductID        Q1         Q2        Q3  \
0         1 2019-03-10           1  11XXRP-P  24924.50   46039.49  41595.55   
1         1 2019-06-04           2  12XXRP-Q  50605.27   44682.74  47928.89   
2         1 2019-01-16           3    13XXRP  13740.12   10887.84  11523.47   
3         1 2019-10-18           4    14XXRP  39954.04  d35351.21  36826.95   
4         1 2019-04-11           5  15XXRP-A  32229.38   29620.81  26468.27   
...     ...        ...         ...       ...       ...        ...       ...   
1882     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1883     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1884     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1885     40 2019-09-30          49  4049XXRP      0.00          0      0.00   
1886     40 2019-06-14          43  4093XXRP  31080.68   26600.39  23709.23   

            Q4  
0         0.00  
1         0.00  
2         0.00  
3         0.00  
4         0.00  
...        ...  
1882   3041.41  
1883   3041.41  
1884   3041.41  
1885    -23.99  
1886  22879.37  

[1887 rows x 8 columns]

In [25]:
# filling NaNs that is equal to the previous non-NaN number "Forward Fill"
dataframe_3.fillna(method = 'ffill')

Store       Date  CategoryID ProductID        Q1         Q2        Q3  \
0         1 2019-03-10           1  11XXRP-P  24924.50   46039.49  41595.55   
1         1 2019-06-04           2  12XXRP-Q  50605.27   44682.74  47928.89   
2         1 2019-01-16           3    13XXRP  13740.12   10887.84  11523.47   
3         1 2019-10-18           4    14XXRP  39954.04  d35351.21  36826.95   
4         1 2019-04-11           5  15XXRP-A  32229.38   29620.81  26468.27   
...     ...        ...         ...       ...       ...        ...       ...   
1882     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1883     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1884     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1885     40 2019-09-30          49  4049XXRP      0.00    3556.22   2965.33   
1886     40 2019-06-14          43  4093XXRP  31080.68   26600.39  23709.23   

            Q4  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  
...        ...  
1882   3041.41  
1883   3041.41  
1884   3041.41  
1885    -23.99  
1886  22879.37  

[1887 rows x 8 columns]

In [26]:
# filling NaNs that is equal to the next non-NaN number "Back Fill"
dataframe_3.fillna(method = 'bfill')

Store       Date  CategoryID ProductID        Q1         Q2        Q3  \
0         1 2019-03-10           1  11XXRP-P  24924.50   46039.49  41595.55   
1         1 2019-06-04           2  12XXRP-Q  50605.27   44682.74  47928.89   
2         1 2019-01-16           3    13XXRP  13740.12   10887.84  11523.47   
3         1 2019-10-18           4    14XXRP  39954.04  d35351.21  36826.95   
4         1 2019-04-11           5  15XXRP-A  32229.38   29620.81  26468.27   
...     ...        ...         ...       ...       ...        ...       ...   
1882     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1883     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1884     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1885     40 2019-09-30          49  4049XXRP      0.00   26600.39  23709.23   
1886     40 2019-06-14          43  4093XXRP  31080.68   26600.39  23709.23   

            Q4  
0     15175.52  
1     15175.52  
2     15175.52  
3     15175.52  
4     15175.52  
...        ...  
1882   3041.41  
1883   3041.41  
1884   3041.41  
1885    -23.99  
1886  22879.37  

[1887 rows x 8 columns]

## Duplicate Data

Another data issue that we will solve in this example is to remove duplicate rows from the DataFrame. To idenfity the distinct values in a particular column, we can use the `unique()` function. This will return an array of all the unique column values.

In [27]:
pd.unique(dataframe_3['Store'])

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40], dtype=int64)

Rather than go through each column in the DataFrame to see all the unique values, we can return a count of unique values for each column in the DataFrame using the `nunique()` function.

In [29]:
dataframe_3.nunique()

Store           40
Date           364
CategoryID      50
ProductID     1781
Q1            1713
Q2            1712
Q3            1724
Q4            1707
dtype: int64

Removing duplicate values can be achieved by using the `drop_duplicates` function. Using `drop_duplicates` can be a very efficient way to clean data, especially if no data is supposed to be duplicated. You must be careful when using this function however, as some data is meant to be duplicated on purpose based on the need of the business/situation.

In [23]:
# by default it will drop based on if there is a duplicate for all column values
dataframe_3.drop_duplicates()

Store       Date  CategoryID ProductID        Q1         Q2        Q3  \
0         1 2019-03-10           1  11XXRP-P  24924.50   46039.49  41595.55   
1         1 2019-06-04           2  12XXRP-Q  50605.27   44682.74  47928.89   
2         1 2019-01-16           3    13XXRP  13740.12   10887.84  11523.47   
3         1 2019-10-18           4    14XXRP  39954.04  d35351.21  36826.95   
4         1 2019-04-11           5  15XXRP-A  32229.38   29620.81  26468.27   
...     ...        ...         ...       ...       ...        ...       ...   
1878     40 2019-11-01          41  4041XXRP     46.00         39     10.00   
1879     40 2019-01-07          42  4042XXRP   4241.41     3748.1   4067.06   
1880     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1885     40 2019-09-30          49  4049XXRP      0.00        NaN       NaN   
1886     40 2019-06-14          43  4093XXRP  31080.68   26600.39  23709.23   

            Q4  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  
...        ...  
1878     42.00  
1879   3641.25  
1880   3041.41  
1885    -23.99  
1886  22879.37  

[1883 rows x 8 columns]

In [30]:
# Dropping rows based on specific column values
dataframe_3.drop_duplicates(subset=['CategoryID'])

Store       Date  CategoryID ProductID         Q1         Q2        Q3  \
0         1 2019-03-10           1  11XXRP-P   24924.50   46039.49  41595.55   
1         1 2019-06-04           2  12XXRP-Q   50605.27   44682.74  47928.89   
2         1 2019-01-16           3    13XXRP   13740.12   10887.84  11523.47   
3         1 2019-10-18           4    14XXRP   39954.04  d35351.21  36826.95   
4         1 2019-04-11           5  15XXRP-A   32229.38   29620.81  26468.27   
5         1 2019-05-30           6    16XXRP    5749.03    9135.87   6060.26   
6         1 2019-12-01           7    17XXRP   21084.08   18310.31  19985.20   
7         1 2019-02-22           8  18XXRP-C   40129.01   37334.83  38717.60   
8         1 2019-06-17           9    19XXRP   16930.99   16562.49  15880.85   
9         1 2019-03-03          10   110XXRP   30721.50   31494.77  29634.13   
10        1 2019-03-12          11   111XXRP   24213.18   21760.75  18706.21   
11        1 2019-08-27          12   112XXRP    8449.54    8654.07   9165.98   
12        1 2019-04-22          13   113XXRP   41969.29    36476.4  37857.68   
13        1 2019-05-06          14   114XXRP   19466.91   18129.02  17491.36   
14        1 2019-11-12          16   116XXRP   10217.55   11873.89  13855.54   
15        1 2019-08-16          17   117XXRP   13223.76   13403.66  13485.61   
16        1 2019-08-13          18   118XXRP    4729.50    19006.5  17623.72   
17        1 2019-05-20          19   119XXRP    1947.05    1490.79   1722.17   
18        1 2019-04-01          20   120XXRP    5034.10     3697.7   4719.89   
19        1 2019-11-09          21   121XXRP    8907.63    8735.83   8949.67   
20        1 2019-03-17          22   122XXRP   13623.98   14807.95  12290.16   
21        1 2019-12-11          23   123XXRP   24146.49   22812.41  21381.85   
22        1 2019-08-17          24   124XXRP    8272.90    8050.24   7366.70   
23        1 2019-11-12          25   125XXRP   11609.50   13268.75  10271.25   
24        1 2019-12-19          26   126XXRP   11737.12   10050.92  10547.60   
25        1 2019-10-16          27   127XXRP    2293.00     2339.5   2494.50   
26        1 2019-03-28          28   128XXRP    1085.29    1008.36   1012.09   
27        1 2019-09-13          29   129XXRP    7024.95    7696.61   7966.55   
28        1 2019-07-15          30   130XXRP    5491.00     5328.5   5735.00   
29        1 2019-02-08          31   131XXRP    3455.92    3228.96   2652.00   
30        1 2019-08-13          32   132XXRP   11875.84   14700.59  20275.45   
31        1 2019-07-14          33   133XXRP    8589.77    8483.54   7611.35   
32        1 2019-11-14          34   134XXRP   14350.83   12852.83  13006.50   
33        1 2019-10-17          35   135XXRP    2770.00       2420   2176.00   
34        1 2019-01-08          36   136XXRP     120.00        432    595.00   
35        1 2019-08-28          37   137XXRP    3438.16    3450.68   3203.31   
36        1 2019-01-10          38   138XXRP  115564.35   94136.35  98672.59   
37        1 2019-03-11          40   140XXRP   66780.63   60828.25  68569.53   
38        1 2019-07-14          41   141XXRP    1011.83     924.53   1206.88   
39        1 2019-06-09          42   142XXRP    8366.71    7284.34   8423.87   
40        1 2019-02-14          44   144XXRP    4674.75    4463.33   4401.78   
41        1 2019-07-04          45   145XXRP      37.44      34.47     22.94   
42        1 2019-01-07          46   146XXRP   20837.77   18826.47  22414.94   
43        1 2019-08-06          47   147XXRP      70.00        NaN   -863.00   
44        1 2019-05-31          48   148XXRP     814.00        247    354.00   
45        1 2019-07-03          49   149XXRP   11501.46   16139.05  16736.17   
46        1 2019-09-26          43   193XXRP   64494.87   70202.02  62581.64   
84        2 2019-08-13          39   239XXRP       0.00        NaN      0.78   
469      10 2019-10-01          50  1050XXRP    2350.00       3704   4620.00   
1611  

## Error Data

Another common cleaning method is to identify and clean errors in our dataset. Consider the first few lines of the DataFrame in the `ProductID` column.

In [45]:
dataframe_3

Store       Date  CategoryID ProductID        Q1         Q2        Q3  \
0         1 2019-03-10           1    11XXRP  24924.50   46039.49  41595.55   
1         1 2019-06-04           2    12XXRP  50605.27   44682.74  47928.89   
2         1 2019-01-16           3    13XXRP  13740.12   10887.84  11523.47   
3         1 2019-10-18           4    14XXRP  39954.04  d35351.21  36826.95   
4         1 2019-04-11           5    15XXRP  32229.38   29620.81  26468.27   
...     ...        ...         ...       ...       ...        ...       ...   
1882     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1883     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1884     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1885     40 2019-09-30          49  4049XXRP      0.00        NaN       NaN   
1886     40 2019-06-14          43  4093XXRP  31080.68   26600.39  23709.23   

            Q4 SpecialID  
0          NaN         P  
1          NaN         Q  
2          NaN      None  
3          NaN      None  
4          NaN         A  
...        ...       ...  
1882   3041.41      None  
1883   3041.41      None  
1884   3041.41      None  
1885    -23.99      None  
1886  22879.37      None  

[1887 rows x 9 columns]

We can see that in the `ProductID` column, that some values have a suffix starting with `-`, which denotes that this product falls under a special category, whereas the ones without are regular products. However, the information that indicates that it is a special item should be in a different column, separated from the Product ID.

To do this, we can use the `str.split()` function to slice and separate values based on the values before and after the delimiter. The delimter in this instance would be `-`.

In [46]:
# Note - Without instantiating the function to the actual columns, no change is made on the actual DataFrame 
dataframe_3['ProductID'].str.split('-', expand = True)

0
0       11XXRP
1       12XXRP
2       13XXRP
3       14XXRP
4       15XXRP
...        ...
1882  4044XXRP
1883  4044XXRP
1884  4044XXRP
1885  4049XXRP
1886  4093XXRP

[1887 rows x 1 columns]

In [33]:
# Instantiating the result above to the the DataFrame. Column 0 replaces 'ProductID', column 1 replaces 'SpecialID'.
dataframe_3[['ProductID', 'SpecialID']] = dataframe_3['ProductID'].str.split('-', expand=True)
dataframe_3

Store       Date  CategoryID ProductID        Q1         Q2        Q3  \
0         1 2019-03-10           1    11XXRP  24924.50   46039.49  41595.55   
1         1 2019-06-04           2    12XXRP  50605.27   44682.74  47928.89   
2         1 2019-01-16           3    13XXRP  13740.12   10887.84  11523.47   
3         1 2019-10-18           4    14XXRP  39954.04  d35351.21  36826.95   
4         1 2019-04-11           5    15XXRP  32229.38   29620.81  26468.27   
...     ...        ...         ...       ...       ...        ...       ...   
1882     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1883     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1884     40 2019-03-24          44  4044XXRP   3280.00    3556.22   2965.33   
1885     40 2019-09-30          49  4049XXRP      0.00        NaN       NaN   
1886     40 2019-06-14          43  4093XXRP  31080.68   26600.39  23709.23   

            Q4 SpecialID  
0          NaN         P  
1          NaN         Q  
2          NaN      None  
3          NaN      None  
4          NaN         A  
...        ...       ...  
1882   3041.41      None  
1883   3041.41      None  
1884   3041.41      None  
1885    -23.99      None  
1886  22879.37      None  

[1887 rows x 9 columns]

## Export Data

Pandas can also export the data once we have finished cleaning it. We can use the `to_csv()` function to do so, and the file will be stored in the same directory as where this notebook is in.

In [34]:
dataframe_3.to_csv('export.csv')